File: app/src/main/java/com/cesarwillymc/jpmorgantest/data/sources/search/local/SearchLocalDataSourceImpl.java
package com.cesarwillymc.jpmorgantest.data.sources.search.local;

import android.content.SharedPreferences;

import androidx.annotation.NonNull;

import com.cesarwillymc.jpmorgantest.util.constants.ConstantsKt;
import com.cesarwillymc.jpmorgantest.util.state.Result;

import javax.inject.Inject;

import kotlin.Unit;
import kotlin.jvm.JvmField;


/**
 * Created by Cesar Canaza on 10/4/23.
 * cesarwilly.mc@gmail.com
 * <p>
 * IOWA, United States.
 */
public class SearchLocalDataSourceImpl implements SearchLocalDataSource {
    private final SharedPreferences sharedPreferences;

    @Inject
    public SearchLocalDataSourceImpl(SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    @NonNull
    @Override
    public Result<Unit> saveQuery(@NonNull String value) {
        try {
            SharedPreferences.Editor editor = sharedPreferences.edit();
            if (value.isEmpty()) {
                editor.remove(RECENTLY_SEARCHED);
            } else {
                editor.putString(RECENTLY_SEARCHED, value);
            }
            editor.apply();
            return new Result.Success<>(null);
        } catch (Exception e) {
            return new Result.Error(e);
        }
    }

    @NonNull
    @Override
    public Result<String> recentlySearched() {
        try {
            String result = sharedPreferences.getString(
                    RECENTLY_SEARCHED,
                    ConstantsKt.EMPTY_STRING
            );
            if (result.isBlank()) {
                return new Result.Error(new Exception());
            } else {
                return new Result.Success<>(result);
            }
        } catch (Exception e) {
            return new Result.Error(e);
        }
    }

    @JvmField
    static String RECENTLY_SEARCHED = "recently_searched";
}

File: app/src/main/java/com/cesarwillymc/jpmorgantest/data/sources/search/local/SearchLocalDataSourceImpl.kt
package com.cesarwillymc.jpmorgantest.data.sources.search.local

import android.content.SharedPreferences
import androidx.core.content.edit
import com.cesarwillymc.jpmorgantest.util.constants.EMPTY_STRING
import com.cesarwillymc.jpmorgantest.util.state.Result
import javax.inject.Inject

/**
 * Created by Cesar Canaza on 10/3/23.
 * cesarwilly.mc@gmail.com
 *
 * IOWA, United States.
 */
class SearchLocalDataSourceImpl @Inject constructor(
    private val sharedPreferences: SharedPreferences
) : SearchLocalDataSource {
    override fun saveQuery(value: String): Result<Unit> {
        return try {
            sharedPreferences.edit {
                if (value.isEmpty()) {
                    remove(RECENTLY_SEARCHED)
                } else {
                    putString(RECENTLY_SEARCHED, value)
                }
            }
            Result.Success(Unit)
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    override fun recentlySearched(): Result<String> {
        return try {
            return Result.Success(
                sharedPreferences.getString(
                    RECENTLY_SEARCHED,
                    EMPTY_STRING
                ).orEmpty()
            )
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    companion object {
        const val RECENTLY_SEARCHED = "recently_searched"
    }
}
